File: module-server/src/main/java/de/twomartens/wahlrecht/mapper/v1/VotingResultMapper.java
package de.twomartens.wahlrecht.mapper.v1;

import de.twomartens.wahlrecht.model.internal.NominationId;
import de.twomartens.wahlrecht.model.internal.VotingResult;
import org.mapstruct.CollectionMappingStrategy;
import org.mapstruct.Mapper;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.ReportingPolicy;

@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface VotingResultMapper {

  default de.twomartens.wahlrecht.model.dto.v1.VotingResult mapToExternal(VotingResult result) {
    return new de.twomartens.wahlrecht.model.dto.v1.VotingResult(
        result.getNominationId().electionName(),
        result.getNominationId().partyAbbreviation(),
        result.getNominationId().name(),
        result.getVotesOnNomination(),
        result.getVotesThroughHealing(),
        result.getVotesPerPosition()
    );
  }

  default VotingResult mapToInternal(de.twomartens.wahlrecht.model.dto.v1.VotingResult result) {
    return new VotingResult(
        new NominationId(result.electionName(), result.partyAbbreviation(),
            result.nominationName()),
        result.votesOnNomination(),
        result.votesThroughHealing(),
        result.votesPerPosition()
    );
  }
}

File: module-server/src/main/kotlin/de/twomartens/wahlrecht/mapper/v1/VotingResultMapper.kt
package de.twomartens.wahlrecht.mapper.v1

import de.twomartens.wahlrecht.model.internal.NominationId
import de.twomartens.wahlrecht.model.internal.VotingResult
import org.mapstruct.CollectionMappingStrategy
import org.mapstruct.Mapper
import org.mapstruct.NullValueCheckStrategy
import org.mapstruct.ReportingPolicy
import org.springframework.lang.NonNull

@Mapper(
    collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE
)
interface VotingResultMapper {
    fun mapToExternal(result: VotingResult): de.twomartens.wahlrecht.model.dto.v1.VotingResult {
        return de.twomartens.wahlrecht.model.dto.v1.VotingResult(
            electionName = result.nominationId.electionName,
            partyAbbreviation = result.nominationId.partyAbbreviation,
            nominationName = result.nominationId.name,
            votesOnNomination = result.votesOnNomination,
            votesThroughHealing = result.votesThroughHealing,
            votesPerPosition = result.votesPerPosition
        )
    }

    fun mapToInternal(@NonNull result: de.twomartens.wahlrecht.model.dto.v1.VotingResult): VotingResult {
        return VotingResult(
            nominationId = NominationId(
                electionName = result.electionName,
                partyAbbreviation = result.partyAbbreviation,
                name = result.nominationName
            ),
            votesOnNomination = result.votesOnNomination,
            votesThroughHealing = result.votesThroughHealing,
            votesPerPosition = result.votesPerPosition
        )
    }
}